src/linker/utilities/spark_utils.py

```diff
@@ -1,65 +1,72 @@
-import atexit
-import os
 import tempfile
 from pathlib import Path
 from time import sleep
-from typing import List, TextIO
+from typing import Any, Dict, List, Optional, TextIO, Tuple
 
 from loguru import logger
-from numpy import diag
 
-from linker.configuration import Config
 from linker.utilities.paths import CONTAINER_DIR
 from linker.utilities.slurm_utils import get_slurm_drmaa, submit_spark_cluster_job
 
 
 def build_cluster(
-    config: Config,
+    session: Optional["drmaa.Session"],
+    resources: Dict[str, Any],
+    step_id: str,
     results_dir: Path,
     diagnostics_dir: Path,
     input_data: List[Path],
-    preserve_logs: bool = False,
-) -> str:
-    """Builds a Spark cluster. Main function for the `build-spark-cluster` command.
+) -> Tuple[str, str]:
+    """Builds a Spark cluster.
 
     Args:
-        config: Config object.
-        preserve_logs: Whether to preserve logs.
+        session: DRMAA session.
+        resources: Resources for the Spark cluster.
+        step_id: Step ID for naming jobs.
+        results_dir: Results directory.
+        diagnostics_dir: Diagnostics directory.
+        input_data: Input data.
 
     Returns:
         spark_master_url: Spark master URL.
     """
     drmaa = get_slurm_drmaa()
-    session = drmaa.Session()
-    session.initialize()
+    if not session:
+        stop_session = True
```

Comment:
What is this doing?

Reply:
Oh...I think this is actually something we don't need at the moment but will need later.

The idea is that if a drmaa session is passed in (i.e. you're running a slurm batch job as opposed to local), then this will use that session to spin up the spark cluster but then NOT kill it after running the spark implementation (b/c we need the session to persist for the entire pipeline).

If running locally, then we do NOT yet have a drmaa session and so we start one for the sole purpose of spinning up the spark cluster. Then, when that implementation is done running, we can kill the session.

But since we've punted on actually spinning up local spark clusters (and are instead relying temporarily on the implementation itself doing that within the container), then I think this isn't strictly necessary.

I'll test.

Reply:
Yeah I don't think this is necessary -- even in the future, the only time we should use Slurm to start the Spark cluster is if we are using Slurm to start the implementations as well.